cylc/flow/ws_data_mgr.py

```diff
@@ -0,0 +1,393 @@
+                    copy(state).replace('-', '_'), count)
+
+        for key, value in (
+                ('oldest_cycle_point', self.schd.pool.get_min_point()),
+                ('newest_cycle_point', self.schd.pool.get_max_point()),
+                ('newest_runahead_cycle_point',
+                    self.schd.pool.get_max_point_runahead())):
+            if value:
+                setattr(workflow, key, str(value))
+            else:
+                setattr(workflow, key, '')
+
+        if get_utc_mode():
+            time_zone_info = TIME_ZONE_UTC_INFO
+        else:
+            time_zone_info = TIME_ZONE_LOCAL_INFO
+        for key, val in time_zone_info.items():
+            setattr(workflow.time_zone_info, key, val)
+
+        workflow.last_updated = update_time
+        workflow.run_mode = self.schd.run_mode
+        workflow.cycling_mode = config.cfg['scheduling']['cycling mode']
+        workflow.states.extend(list(set(all_states)).sort())
+        workflow.reloading = self.schd.pool.do_reload
+        workflow.workflow_log_dir = self.schd.suite_log_dir
+        workflow.job_log_names.extend(list(JOB_LOG_OPTS.values()))
+        workflow.ns_defn_order.extend(config.ns_defn_order)
+
+        # Construct a workflow status string for use by monitoring clients.
+        if self.schd.pool.is_held:
+            status_string = SUITE_STATUS_HELD
+        elif self.schd.stop_mode is not None:
+            status_string = SUITE_STATUS_STOPPING
+        elif self.schd.pool.hold_point:
+            status_string = (
+                SUITE_STATUS_RUNNING_TO_HOLD %
+                self.schd.pool.hold_point)
+        elif self.schd.stop_point:
+            status_string = (
+                SUITE_STATUS_RUNNING_TO_STOP %
+                self.schd.stop_point)
+        elif self.schd.stop_clock_time is not None:
+            status_string = (
+                SUITE_STATUS_RUNNING_TO_STOP %
+                self.schd.stop_clock_time_string)
+        elif self.schd.stop_task:
+            status_string = (
+                SUITE_STATUS_RUNNING_TO_STOP %
+                self.schd.stop_task)
+        elif self.schd.final_point:
+            status_string = (
+                SUITE_STATUS_RUNNING_TO_STOP %
+                self.schd.final_point)
+        else:
+            status_string = SUITE_STATUS_RUNNING
+        workflow.status = status_string
+
+        workflow.tasks.extend([t.id for t in tasks.values()])
+        workflow.families.extend([f.id for f in families.values()])
+
+        # Generate ungrouped edges
+        try:
+            graph_edges = config.get_graph_edges(
+                workflow.oldest_cycle_point,
+                workflow.newest_cycle_point,
+                group_nodes=None, ungroup_nodes=None,
+                ungroup_recursive=False, group_all=False
+            )
+        except TypeError:
+            graph_edges = []
+
+        if isinstance(graph_edges, list) and graph_edges != []:
```

Comment:
`if isinstance(graph_edges, list) and graph_edges:` or `and len(graph_edges)`. This way avoids the creation of an empty list for comparison (in theory, though there could be some extra optimization in Python/bytecode/etc, but I think it's recommended to use `not`, `len`, etc).

Reply:
N/A